processing/src/main/java/org/apache/druid/math/expr/ExprEval.java

```diff
@@ -1314,38 +1337,87 @@ public ExpressionType type()
     @Override
     public boolean isNumericNull()
     {
-      return false;
+      if (ExpressionType.NESTED_DATA.equals(expressionType)) {
+        computeNumber();
+        return number == null;
+      }
+      return true;
     }
 
     @Override
     public int asInt()
     {
+      if (ExpressionType.NESTED_DATA.equals(expressionType)) {
+        computeNumber();
+        if (number != null) {
+          return number.intValue();
+        }
+      }
       return 0;
     }
 
     @Override
     public long asLong()
     {
+      if (ExpressionType.NESTED_DATA.equals(expressionType)) {
+        computeNumber();
+        if (number != null) {
+          return number.longValue();
+        }
+      }
       return 0;
     }
 
     @Override
     public double asDouble()
     {
+      if (ExpressionType.NESTED_DATA.equals(expressionType)) {
+        computeNumber();
+        if (number != null) {
+          return number.doubleValue();
+        }
+      }
       return 0;
     }
 
     @Override
     public boolean asBoolean()
     {
+      if (ExpressionType.NESTED_DATA.equals(expressionType)) {
+        Object val = StructuredData.unwrap(value);
+        if (val != null) {
+          return Evals.objectAsBoolean(val);
+        }
+      }
       return false;
     }
 
+    private void computeNumber()
+    {
+      if (number == null && value != null) {
+        Object val = StructuredData.unwrap(value);
+        if (val instanceof Number) {
+          number = (Number) val;
+        } else if (val instanceof Boolean) {
+          number = Evals.asLong((Boolean) val);
+        } else if (val instanceof String) {
+          number = ExprEval.computeNumber((String) val);
+        }
+      }
+    }
+
     @Nullable
     @Override
     public Object[] asArray()
     {
-      return new Object[0];
+      if (ExpressionType.NESTED_DATA.equals(expressionType)) {
```

Comment:
Just a thought. There's so much of this logic that perhaps it would be better to have `NestedDataExprEval` be its own subclass of `ExprEval`, rather than embedding all this conditional logic into `ComplexExprEval`.

`ComplexExprEval` is a private class anyway, so this could be done transparently to callers outside this file.

Reply:
good call, done